### PR TITLE
Refactor the Jenkins plugin

### PR DESCRIPTION
**Pull request by sameagen-MW**
# Intro
Here is the first draft of the refactor, but there are a few action items that I would like to follow up on with various people. I'll follow up more on this at the end of the review request, but first I'll outline the new architecture:

# Architecture
All logic related to interacting with the filesystem or spawning a new process has been moved into the `MatlabCommandRunner` class.

This class is owned by composition by the three new action classes, `RunMatlabCommandAction`, `RunMatlabBuildAction`, and `RunMatlabTestsAction`. These action classes form the shared logic that is used by both freestyle and pipeline build steps.

the `RunMatlab*Builder` classes have been moved into a new `com.mathworks.ci.freestyle` subnamespace, and associated classes used for UI elements have been moved into a `com.mathworks.ci.freestyle.options` namespace. To facilitate backwards compatibility XStream aliases have been added to all freestyle builders.

The `RunMatlab*Step` and `Matlab*StepExecution` classes have been moved to a new `com.mathworks.ci.pipeline` namespace. Because the step configuration serialization is unaffected by these changes, no aliases are required.

I have moved the existing tests to the `src/test/java/integ` subfolder, and added a new `src/test/java/unit` folder to hold the new unit tests. The new unit tests aim to be much smaller in scope than the existing tests and thus hopefully will not increase the runtime of the already quite long running suite.

# Future Actions

I modified the existing tests the least possible amout while still attempting to preserve the intention of the test. However, while updating the tests I was unsure of whether a decent number of the tests were still useful. I believe that there is a lot more that can be done to update the existing test suite and potentially remove a few tests that are no longer adding value. I am happy to take that on when I have some spare time, but would like some additional input from others that have more insight into the purpose of each test. If someone else wants to take on the role of looking at and updating the old integ tests I would also be happy with that.

If I don't hear anything back about testing I'll probably reach out directly to @Vahila and @nbhoski and start working on updating the older set of tests sometime in the next month or two!

For qualification, I am pretty confident that these updates will work for all freshly created jobs, but most of my backwards incompatibility checks were done manually so it might be good to have others bash on that as well to make sure edge cases (like the serialized `SourceFolders` object on a run tests step) still work. 

Additionally, I would love if you could check and make sure that all of the functionality that you added for annotating the log in the run build step still works @nikhilbhoski.

# Overall breakdown of responsibilities

Because this review is kinda massive, I thought it would be helpful to break it up into slightly smaller pieces to give everyone something to primarily focus their efforts toward.

@Vahila - Updates to existing tests, all new test files.
@nikhilbhoski - Updates to `RunMatlab*Step.java`, `Matlab*StepExecution.java`, `RunMatlab*Builder.java` files, `RunMatlabBuildAction.java`. Especially focusing on maintaing backwards compatibility. Updates to existing test files if you have time.
@davidbuzinski - `MatlabCommandRunner.java`, `RunMatlab*Action.java`.

These are just some suggestions, so please feel free to review as much as you want, the more the better! I just want to make sure that everything gets checked out by at least one other person, without anyone feeling overwhelmed by the size of the review. Also please take as much time as you need, I don't think anything here is particularly time sensitive!

Thanks again everybody!

### CODE REVIEW

**Comment by sameagen-MW** [comment]
Have to take off a bit early today, I'll get the spotbugs errors fixed tomorrow!

**Comment by Vahila** [comment]
> # Intro
> Here is the first draft of the refactor, but there are a few action items that I would like to follow up on with various people. I'll follow up more on this at the end of the review request, but first I'll outline the new architecture:
> 
> # Architecture
> All logic related to interacting with the filesystem or spawning a new process has been moved into the `MatlabCommandRunner` class.
> 
> This class is owned by composition by the three new action classes, `RunMatlabCommandAction`, `RunMatlabBuildAction`, and `RunMatlabTestsAction`. These action classes form the shared logic that is used by both freestyle and pipeline build steps.
> 
> the `RunMatlab*Builder` classes have been moved into a new `com.mathworks.ci.freestyle` subnamespace, and associated classes used for UI elements have been moved into a `com.mathworks.ci.freestyle.options` namespace. To facilitate backwards compatibility XStream aliases have been added to all freestyle builders.
> 
> The `RunMatlab*Step` and `Matlab*StepExecution` classes have been moved to a new `com.mathworks.ci.pipeline` namespace. Because the step configuration serialization is unaffected by these changes, no aliases are required.
> 
> I have moved the existing tests to the `src/test/java/integ` subfolder, and added a new `src/test/java/unit` folder to hold the new unit tests. The new unit tests aim to be much smaller in scope than the existing tests and thus hopefully will not increase the runtime of the already quite long running suite.
> 
> # Future Actions
> I modified the existing tests the least possible amout while still attempting to preserve the intention of the test. However, while updating the tests I was unsure of whether a decent number of the tests were still useful. I believe that there is a lot more that can be done to update the existing test suite and potentially remove a few tests that are no longer adding value. I am happy to take that on when I have some spare time, but would like some additional input from others that have more insight into the purpose of each test. If someone else wants to take on the role of looking at and updating the old integ tests I would also be happy with that.
> 
> If I don't hear anything back about testing I'll probably reach out directly to @Vahila and @nbhoski and start working on updating the older set of tests sometime in the next month or two!
> 
> For qualification, I am pretty confident that these updates will work for all freshly created jobs, but most of my backwards incompatibility checks were done manually so it might be good to have others bash on that as well to make sure edge cases (like the serialized `SourceFolders` object on a run tests step) still work.
> 
> Additionally, I would love if you could check and make sure that all of the functionality that you added for annotating the log in the run build step still works @nikhilbhoski.
> 
> # Overall breakdown of responsibilities
> Because this review is kinda massive, I thought it would be helpful to break it up into slightly smaller pieces to give everyone something to primarily focus their efforts toward.
> 
> @Vahila - Updates to existing tests, all new test files. @nikhilbhoski - Updates to `RunMatlab*Step.java`, `Matlab*StepExecution.java`, `RunMatlab*Builder.java` files, `RunMatlabBuildAction.java`. Especially focusing on maintaing backwards compatibility. Updates to existing test files if you have time. @davidbuzinski - `MatlabCommandRunner.java`, `RunMatlab*Action.java`.
> 
> These are just some suggestions, so please feel free to review as much as you want, the more the better! I just want to make sure that everything gets checked out by at least one other person, without anyone feeling overwhelmed by the size of the review. Also please take as much time as you need, I don't think anything here is particularly time sensitive!
> 
> Thanks again everybody!

Hi @sameagen-MW, I can look into the testing aspect and see what tests can be removed/added/updated.

**Comment by sameagen-MW** [comment]
> > # Intro
> > Here is the first draft of the refactor, but there are a few action items that I would like to follow up on with various people. I'll follow up more on this at the end of the review request, but first I'll outline the new architecture:
> > # Architecture
> > All logic related to interacting with the filesystem or spawning a new process has been moved into the `MatlabCommandRunner` class.
> > This class is owned by composition by the three new action classes, `RunMatlabCommandAction`, `RunMatlabBuildAction`, and `RunMatlabTestsAction`. These action classes form the shared logic that is used by both freestyle and pipeline build steps.
> > the `RunMatlab*Builder` classes have been moved into a new `com.mathworks.ci.freestyle` subnamespace, and associated classes used for UI elements have been moved into a `com.mathworks.ci.freestyle.options` namespace. To facilitate backwards compatibility XStream aliases have been added to all freestyle builders.
> > The `RunMatlab*Step` and `Matlab*StepExecution` classes have been moved to a new `com.mathworks.ci.pipeline` namespace. Because the step configuration serialization is unaffected by these changes, no aliases are required.
> > I have moved the existing tests to the `src/test/java/integ` subfolder, and added a new `src/test/java/unit` folder to hold the new unit tests. The new unit tests aim to be much smaller in scope than the existing tests and thus hopefully will not increase the runtime of the already quite long running suite.
> > # Future Actions
> > I modified the existing tests the least possible amout while still attempting to preserve the intention of the test. However, while updating the tests I was unsure of whether a decent number of the tests were still useful. I believe that there is a lot more that can be done to update the existing test suite and potentially remove a few tests that are no longer adding value. I am happy to take that on when I have some spare time, but would like some additional input from others that have more insight into the purpose of each test. If someone else wants to take on the role of looking at and updating the old integ tests I would also be happy with that.
> > If I don't hear anything back about testing I'll probably reach out directly to @Vahila and @nbhoski and start working on updating the older set of tests sometime in the next month or two!
> > For qualification, I am pretty confident that these updates will work for all freshly created jobs, but most of my backwards incompatibility checks were done manually so it might be good to have others bash on that as well to make sure edge cases (like the serialized `SourceFolders` object on a run tests step) still work.
> > Additionally, I would love if you could check and make sure that all of the functionality that you added for annotating the log in the run build step still works @nikhilbhoski.
> > # Overall breakdown of responsibilities
> > Because this review is kinda massive, I thought it would be helpful to break it up into slightly smaller pieces to give everyone something to primarily focus their efforts toward.
> > @Vahila - Updates to existing tests, all new test files. @nikhilbhoski - Updates to `RunMatlab*Step.java`, `Matlab*StepExecution.java`, `RunMatlab*Builder.java` files, `RunMatlabBuildAction.java`. Especially focusing on maintaing backwards compatibility. Updates to existing test files if you have time. @davidbuzinski - `MatlabCommandRunner.java`, `RunMatlab*Action.java`.
> > These are just some suggestions, so please feel free to review as much as you want, the more the better! I just want to make sure that everything gets checked out by at least one other person, without anyone feeling overwhelmed by the size of the review. Also please take as much time as you need, I don't think anything here is particularly time sensitive!
> > Thanks again everybody!
> 
> Hi @sameagen-MW, I can look into the testing aspect and see what tests can be removed/added/updated.

Thanks @Vahila!

File: src/main/java/com/mathworks/ci/actions/RunMatlabTestsAction.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        final List<String> inputArgsList = new ArrayList<String>();
+
+        inputArgsList.add("'Test'");
+
+        // Prepare source and test folder lists
+        String sourceFolders = null;
+        if (this.params.getSourceFolder() != null) {
+            sourceFolders = this.params.getSourceFolder().size() == 0
+                ? null
+                : Utilities.getCellArrayFrmList(this.params.getSourceFolder());
```
Comment (davidbuzinski) [inline src/main/java/com/mathworks/ci/actions/RunMatlabTestsAction.java:73]: Seems a little funny to drop the "o" in `From` in an already long method name 😆

File: src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        // Prepare the executable
+        FilePath exePath = prepareRunnerExecutable();
+
+        // Create the script file
+        FilePath scriptFile = createFileWithContent(command);
+        String cmd = "setenv('MW_ORIG_WORKING_FOLDER', cd('"
+            + this.tempFolder.getRemote()
+            + "'));"
+            + scriptFile.getBaseName();
+
```
Comment (davidbuzinski) [inline src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java:80]: Oh I see, this fixes the `-sd` thing here, but we need `addPath` only for `genscript` and plugins. Disregard the above commend 👍

File: src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+     */
+    public void addEnvironmentVariable(String key, String value) {
+        additionalEnvVars.put(key, value);
+    }
+
+    public EnvVars getEnvVars() {
+        EnvVars env = new EnvVars(this.params.getEnvVars());
+        env.putAll(additionalEnvVars); 
+        return env;
+    }
```
Comment (davidbuzinski) [inline src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java:134]: I sort of don't understand why we're doing this. It looks like we wrap environment variables twice: once in our parameters class definition and once in our runner class definition. Is this mostly for testing? It makes things a little confusing for me below when we create the script.

File: src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+     * @param content string that represents the content of the file.
+     * @return the FilePath to the script file that is created.
+     */
+    protected FilePath createFileWithContent(String content) throws IOException, InterruptedException {
+        String fileName = "script_" + RandomStringUtils.randomAlphanumeric(8) + ".m";
+        FilePath scriptFile = new FilePath(this.tempFolder, fileName);
+
+        String expandedContent = getEnvVars().expand(content);
+        String finalContent = "cd(getenv('MW_ORIG_WORKING_FOLDER'));\n"
+            + expandedContent;
```
Comment (davidbuzinski) [inline src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java:171]: I don't understand the expanded content here. Can you help give an example of what it will look like with and without environment variables?

File: src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                    params.getTaskListener().getLogger(), 
+                    params.getBuild().getCharset()),
+                params);
+    }
+
+    public void run() throws IOException, InterruptedException, MatlabExecutionException {
+        // Copy plugins and override default plugins function
+        runner.copyFileToTempFolder(DEFAULT_PLUGIN, DEFAULT_PLUGIN);
+        runner.copyFileToTempFolder(BUILD_REPORT_PLUGIN, BUILD_REPORT_PLUGIN);
+        runner.copyFileToTempFolder(TASK_RUN_PROGRESS_PLUGIN, TASK_RUN_PROGRESS_PLUGIN);
```
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java:51]: Great suggestion! I'll switch that.

File: src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        FilePath jsonFile = new FilePath(params.getWorkspace(), ".matlab/buildArtifact.json");
+        if (jsonFile.exists()) {
+            FilePath rootLocation = new FilePath(
+                    new File(
+                        build.getRootDir()
+                        .getAbsolutePath()
+                        + "/buildArtifact.json"));
+            jsonFile.copyTo(rootLocation);
+            jsonFile.delete();
+            build.addAction(new BuildArtifactAction(build, this.params.getWorkspace()));
```
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java:97]: Yeah, this is the same behavior that it has in the most recent release which added the build artifact + results action. I think that we can probably rename the file without causing any issues so I will go ahead and update that.

My understanding is similar to yours, since it's using this interface:
https://javadoc.jenkins-ci.org/hudson/model/PersistenceRoot.html

I'll double check though to make sure that no build reports are getting overwritten by confirming where the build root is in practice.

File: src/main/java/com/mathworks/ci/actions/RunMatlabTestsAction.java (hunk 1 above)
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/actions/RunMatlabTestsAction.java:73]: Oh yeah, I didn't end up touching the `Utilities` class but I can make a quick pass over it and normalize it to be more similar to the rest of the updated codebase.

File: src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java (hunk 3 above)
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java:134]: Great question! The simple answer is that I hit sporadic concurrency issues in the pipeline steps when I tried to rely on one set of environment variables stored in the parameters. The issue was due to one piece of the code trying to add environment variables while another location was reading the environment variables field.

By having a separate mutable set of environment variables owned solely by the `MatlabCommandRunner` and then joining those with the "master" copy stored inside of the parameters object and returning it as a new object we can avoid any concurrent access violations.

For a little bit I was serializing the parameters object rather than constructing it within the `StepExecution` class, so it's likely that concurrent access is no longer an issue (I think it's likely the serialization process was what was reading the parameters object's fields while the build action was adding an environment variable), but I still like keeping the parameter objects totally immutable to make it easier for contributors in the future to avoid running into the same issues.

File: src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java (hunk 4 above)
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java:171]: Yeah!

In the following example, the user entered the literal string `"${TASKS}"` into the tasks UI textbox, and nothing into the build options textbox.

The user also has set the environment variable `TASKS` to be `clean lint test` on their build agent.

Result:
```
contentBeforeExpansion = "buildtool ${TASKS}"
contentAfterExpansion = "buildtool clean lint test"
```

File: src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        } catch (Exception e) {
+            this.params.getTaskListener().getLogger()
+                .println(e.getMessage());
+            throw(e);
+        } finally {
+            annotator.forceEol();
+        }
+
+        // Handle build result
+        Run<?,?> build = this.params.getBuild();
```
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java:87]: It looks like that's the idiomatic way to use that structure, I don't see any examples of people actually filling in the fields: https://github.com/search?q=%2F%28%3F-i%29+Run%3C%5B%5E%5C%2F%5D%2F+language%3AJava+&type=code

File: src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java:51]: Hmm, this might actually be a little bit trickier than I thought before. It doesn't seem like there's any great ways to copy a folder out of the jar resources, so I think the change would probably require us to zip the entire folder up and then unzip it in that location.

I'm not sure that's worth it, as I think the change to zipping and unzipping would take up more space than just having the files listed out (unless we start adding a bunch more plugins.) What's your opinion on leaving it the way it is for now?

[the remark this answers is not in the document]
Comment (davidbuzinski) [reply src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java:51]: Yeah leaving it this way is okay.

File: src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java:97]: No flag however this might introduce compatibility issues with the files which are already archived under root. Considering users have used the recent version of plugin and archives few build results. Any plans on how we will take care of it? or we are safe enough to assume that the impact may not be widers as the release is very recent.

File: src/main/java/com/mathworks/ci/freestyle/RunMatlabCommandBuilder.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+import com.mathworks.ci.Message;
+import com.mathworks.ci.parameters.RunActionParameters;
+import com.mathworks.ci.actions.MatlabActionFactory;
+import com.mathworks.ci.actions.RunMatlabCommandAction;
+import com.mathworks.ci.freestyle.options.StartupOptions;
+
+public class RunMatlabCommandBuilder extends Builder implements SimpleBuildStep {
+    // Deprecated
+    private transient int buildResult;
```
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/freestyle/RunMatlabCommandBuilder.java:39]: Unfortunately removing a given field rather than just setting it to transient can sometimes mess up the deserialization of past builds. For backwards compatibility the recommended approach is to instead switch the field to transient so it will no longer be serialized in future jobs but is still compatible with previous jobs.

File: src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        FilePath jsonFile = new FilePath(params.getWorkspace(), ".matlab/buildArtifact.json");
```
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java:88]: Yep, I'll get that fixed.

File: src/main/java/com/mathworks/ci/freestyle/RunMatlabTestsBuilder.java [hunk 10]
```diff
@@ -194,14 +203,33 @@ public String getPdfReportFilePath() {
     public SelectByTag getSelectByTag() {
     	return this.selectByTag;
     }
+
+    public String getSelectByTagPath() {
```
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/freestyle/RunMatlabTestsBuilder.java:207]: Makes sense to me!

File: src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+                "MW_MATLAB_BUILDTOOL_DEFAULT_PLUGINS_FCN_OVERRIDE",
+                "ciplugins.jenkins.getDefaultPlugins");
+
+        // Redirect output to the build annotator
+        runner.redirectStdOut(annotator);
+
+        // Prepare the build tool command
+        // TODO: Devise better solution then prepending the command
+        //   here.
+        String command = "addpath('"
```
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java:64]: I believe that the idea (both here and for the test action) is that we need some extra pieces of code available rather than just the script that contains the command that we're calling. In this case we need the build plugins to be on the path, and for the test action we need genscript.

In particular, we want the script that the user is calling to run in the directory that we launched MATLAB in, so we change back to that directory as the first line of the scratch script. However, after we `cd` back, we still need the function which overrides the default plugins to be on the path.

If you have any suggestions for other ways to resolve this definitely let me know!

File: src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/actions/RunMatlabBuildAction.java:97]: Hmm, good point. What are your thoughts @davidbuzinski ? I'm not sure that renaming it is worth the risk of an incompatibility but it seems like if we are going to rename it the sooner we do the better. 

I think I'd prefer to avoid supporting both artifacts named `buildArtifact.json` and `matlabBuildArtifact.json` which is my first thought as to how we would avoid the incompatibility.

**Review by davidbuzinski — APPROVED** [review]

File: src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    }
+
+    /**
+     * Copies platform specific runner file into the temporary folder.
+     *
+     * @return the FilePath to the runner executable
+     */
+    protected FilePath prepareRunnerExecutable() throws IOException, InterruptedException {
+        Launcher launcher = this.params.getLauncher();
+        if (launcher.isUnix()) {
```
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java:188]: I believe that this is necessary to get the correct executable for the platform the plugin is running on. We'll need to check the operating system and differentiate between Mac and Linux at some point, and this seemed like the natural place to do so.

Getting the destination folder name is far less important than getting the correct source folder location. In fact I believe that the destination location is the same for both branches, we just copy a different executable to that destination.

This is also identical to how we're differentiating between Mac and Linux in the current iteration of the plugin. If you have another suggestion that you believe would be an improvement definitely let me know, because ideally we wouldn't need to spawn an extra process.

File: src/main/java/com/mathworks/ci/pipeline/MatlabBuildStepExecution.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        this.step = step;
+    }
+    
+    public MatlabBuildStepExecution(StepContext ctx, RunMatlabBuildStep step) throws IOException, InterruptedException {
+        this(new MatlabActionFactory(), ctx, step);
+    }
+
+    @Override
+    public Void run() throws Exception {
+        BuildActionParameters params = new BuildActionParameters(
```
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/pipeline/MatlabBuildStepExecution.java:38]: Yes, this is an issue when you use scripted pipeline especially with node {} block. although the updated document mentions that it allocates Workspace I did not see that happening. I guess the script throws error if it does not find Workspace. so I am ok if works fine with the scripted pipeline.

[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/pipeline/MatlabBuildStepExecution.java:38]: I'm still a little bit confused as to what you mean. Do you mean that in the past the `node` step has not allocated a workspace in your experience but it seems like they have updated it to allocate one?

If so, judging from [this](https://groups.google.com/g/jenkinsci-users/c/513qLiYlXHc) thread, my guess would be that that was a bug rather than an intended feature. Probably worth ensuring that we can fail gracefully in that case however.

File: src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java:188]: is `run-matlab-command` different for MAc & linux ?

[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java:188]: Yep!

image

[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java:188]: Ahh, makes sense in that case.

File: src/main/java/com/mathworks/ci/pipeline/MatlabBuildStepExecution.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/pipeline/MatlabBuildStepExecution.java:38]: Yes same error but i had seen it if I use  node{}

[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/pipeline/MatlabBuildStepExecution.java:38]: Gotcha! It seems like other popular plugins are going ahead and creating the workspace if it doesn't exist, so I will make the fix that you recommended off the bat.

I'll probably add it to `MatlabCommandRunner` so that the behavior is the same across all steps if that makes sense?

File: src/main/java/com/mathworks/ci/freestyle/RunMatlabBuildBuilder.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+import jenkins.tasks.SimpleBuildStep;
+import net.sf.json.JSONObject;
+
+import com.mathworks.ci.Message;
+import com.mathworks.ci.actions.MatlabActionFactory;
+import com.mathworks.ci.actions.RunMatlabBuildAction;
+import com.mathworks.ci.parameters.BuildActionParameters;
+import com.mathworks.ci.freestyle.options.*;
+
+public class RunMatlabBuildBuilder extends Builder implements SimpleBuildStep {
```
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/main/java/com/mathworks/ci/freestyle/RunMatlabBuildBuilder.java:35]: Do you know why the `SimpleBuildStep` interface was added to begin with? I can't find much documentation on what exactly it does or what relies upon other objects implementing it, which makes me a bit hesitant to remove it without more info.

[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/freestyle/RunMatlabBuildBuilder.java:35]: initially the idea was to treat the same implementation as pipeline code however later we learnt that this is not very feasible and need a separate implementation. And it remained as is. you can find some info here 
https://groups.google.com/g/jenkinsci-dev/c/hNez69m1ClA/m/HlnZBFd4BAAJ

**Review by nbhoski — APPROVED** [review]
Rest all LGTM

File: src/test/java/unit/com/mathworks/ci/actions/RunMatlabBuildActionTest.java [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    @Mock TaskListener listener;
+    @Mock PrintStream out;
+    @Mock Run build;
+
+    @Mock FilePath tempFolder;
+
+    private boolean setup = false;
+    private RunMatlabBuildAction action;
+
+    @Before
```
Comment (Vahila) [inline src/test/java/unit/com/mathworks/ci/actions/RunMatlabBuildActionTest.java:47]: You can use BeforeClass which will be run once before all the tests. https://junit.org/junit4/javadoc/4.12/org/junit/BeforeClass.html

Same for all other tests.

Comment (sameagen-MW) [reply src/test/java/unit/com/mathworks/ci/actions/RunMatlabBuildActionTest.java:47]: Oh yeah, using `BeforeClass` requires using a `static` method, which would mean that all of the other properties on the class set in `init` also have to be `static`. I generally prefer to stay away from static variables if possible, but in this case I don't think switching the test class properties to be static would have any serious negative (or positivie) impact. Overall though, I still lean slightly towards avoiding it due to some negative past experiences with having to reason about static variables.

Let me know if you feel strongly and I can get that switched over if it's a dealbreaker for you, but using `@Before` instead of `@BeforeClass` was definitely a conscious choice if that changes your input at all. I'd rather be setting encapsulated data on the test class object itself than having to set static data in a static init method.

File: src/main/java/com/mathworks/ci/freestyle/RunMatlabBuildBuilder.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/freestyle/RunMatlabBuildBuilder.java:35]: Sure makes sense if its already deprecated. I dnt have a strong opinioin. We can leave it as is.

File: src/test/java/unit/com/mathworks/ci/actions/RunMatlabBuildActionTest.java (hunk 15 above)
Comment (Vahila) [reply src/test/java/unit/com/mathworks/ci/actions/RunMatlabBuildActionTest.java:47]: I think it's fine to use @Before. Can we add a comment regarding the same?

Comment (sameagen-MW) [reply src/test/java/unit/com/mathworks/ci/actions/RunMatlabBuildActionTest.java:47]: Yep, I'll get a comment added.

File: src/test/java/integ/com/mathworks/ci/BuildArtifactActionTest.java [hunk 16]
```diff
@@ -20,6 +20,8 @@
 import org.junit.Test;
 import org.jvnet.hudson.test.JenkinsRule;
 
+import com.mathworks.ci.freestyle.RunMatlabBuildBuilder;
+
 public class BuildArtifactActionTest {
```
[the remark this answers is not in the document]
Comment (sameagen-MW) [reply src/test/java/integ/com/mathworks/ci/BuildArtifactActionTest.java:25]: Yeah I'd be happy to do that in the future! For now though, because `BuildArtifactAction`'s code hasn't been touched by these changes and none of the tests covering it have been removed I think I'm going to leave that out of scope for this particular PR since it's already on the larger side. 

I want to revisit and update some of the integ tests in the future as well, so I can make that change when I get back to do those tests.

**Review by Vahila — APPROVED** [review]
LGTM!
